src/irmin-pack/unix/io.ml

```diff
@@ -0,0 +1,282 @@
+      let exists = Sys.file_exists path in
+      match exists with
+      | false ->
+          (* just create a file where there wasn't one before *)
+          let fd =
+            Unix.(
+              openfile path
+                [ O_CREAT; O_RDWR; O_EXCL; O_CLOEXEC ]
+                default_create_perm)
+          in
+          Ok { fd; closed = false; readonly = false; path }
+      | true -> (
+          (* file already exists; should we overwrite? *)
+          match overwrite with
+          | true ->
+              (* open the file, with O_TRUNC flag; assumes we have the right permissions
+                 etc *)
+              let fd =
+                Unix.(
+                  openfile path
+                    [ O_RDWR; O_CLOEXEC; O_TRUNC ]
+                    default_create_perm)
+              in
+              Ok { fd; closed = false; readonly = false; path }
+          | false ->
+              (* the file already exists, and we can't overwrite it *)
+              [%log.warn
+                "%s: attempt to create a path %s that already exists \
+                 (overwrite is false)"
+                __FILE__ path];
+              raise
+                Unix.(
+                  Unix_error
+                    ( EEXIST,
+                      (* second component is function name *)
+                      Printf.sprintf "%s: create" __FILE__,
+                      (* third is parameters *)
+                      Printf.sprintf "path=%s; overwrite=%b" path overwrite )))
+    with Unix.Unix_error (e, s1, s2) -> Error (`Io_misc (e, s1, s2))
+
+  let open_ ~path ~readonly : (t, [> open_error ]) result =
+    try
+      let ok = Sys.file_exists path in
+      assert ok;
+      let mode = Unix.(if readonly then O_RDONLY else O_RDWR) in
+      let fd = Unix.(openfile path [ mode; O_CLOEXEC ] default_open_perm) in
+      Ok { fd; closed = false; readonly; path }
+    with Unix.Unix_error (e, s1, s2) -> Error (`Io_misc (e, s1, s2))
+
+  let close t : (unit, [> close_error ]) result =
+    try
+      match t.closed with
+      | true -> Error `Double_close
+      | false ->
+          t.closed <- true;
+          Unix.close t.fd;
+          Ok ()
+    with Unix.Unix_error (e, s1, s2) -> Error (`Io_misc (e, s1, s2))
+
+  let write_string t ~off s : (unit, [> write_error ]) result =
+    try
+      match (t.closed, t.readonly) with
+      | true, _ -> Error `Write_on_closed
+      | _, true -> Error `Ro_not_allowed
+      | _ ->
+          (* really_write and following do not mutate the given buffer, so
+             Bytes.unsafe_of_string is actually safe *)
+          let buf = Bytes.unsafe_of_string s in
+          let () = Util.really_write t.fd off buf 0 (Bytes.length buf) in
+          Ok ()
+    with Unix.Unix_error (e, s1, s2) -> Error (`Io_misc (e, s1, s2))
+
+  (* NOTE copied from write_string above; ensure the code for both functions remains in
+     sync *)
```

Comment:
maybe rewrite `write_string` as a wrapper around `write_exn ` to avoid keeping the two in sync?

Reply:
Yeah, this is likely an improvement.